### PR TITLE
Fix #1348. ctrl+D/U correct position

### LINKED ISSUE #1348: Ctrl-D and Ctrl-U do not update cursor correctly
* Click *thumbs-up* 👍 on this issue if you want it!
* Click *confused* 😕 on this issue if not having it makes VSCodeVim unusable. 

The VSCodeVim team prioritizes issues based on reaction count. 

--------

### What did you do?

Use Ctrl-D and Ctrl-U as movement keys. Also affects Ctrl-f and Ctrl-b.

### What did you expect to happen?

The cursor should retain its relative position on the screen (disregard the edge of file case) during Ctrl-D and Ctrl-U movement.

### What happened instead?

The cursor position is not updated within the file and drops off the visible portion of the screen. On the next successive movement there appears to be some check for an out of bounds cursor which places it to the bottom of the screen regardless of its initial position.

This makes movement with these keys unusable.

### Technical details:

Pretty certain this a regression since VSCode 1.9 which did not have this problem. Doesn't appear to be any recent issues on VSCode itself relating to movement right now so not sure who the cause of this is.

* VSCode Version: 1.10
* VsCodeVim Version: 0.5.3
* OS: Arch Linux + Windows 10

### PR DESCRIPTION
**Pull request by rebornix**
**What this PR does / why we need it**:

We added basic view information to the editor `editor.visibleRanges` in vscode, with which we can do some math to make Ctrl+D/U position correct.

**Which issue(s) this PR fixes**
#1348

**Special notes for your reviewer**:
It requires VSCode 1.22 at least so I updated the vscode dependency. Once we merge this RP and publish, users using VSCode prior to 1.22 will not get the update.

### CODE REVIEW

**Comment by rebornix** [comment]
This PR also demonstrate how we can get proper view information from VSCode API. We can probably improve commands which are not correct or perfect before cc @VSCodeVim/maintainers

File: src/actions/commands/actions.ts [hunk 1]
```diff
@@ -653,13 +653,66 @@ class CommandMoveHalfPageDown extends CommandEditorScroll {
   keys = ['<C-d>'];
   to: EditorScrollDirection = 'down';
   by: EditorScrollByUnit = 'halfPage';
+
+  public async exec(position: Position, vimState: VimState): Promise<VimState> {
+    let lineOffset = 0;
```
Comment (jpoon) [inline src/actions/commands/actions.ts]: I think putting this in the base class `CommandEditorScroll` is a better place for this?

Comment (rebornix) [reply src/actions/commands/actions.ts]: I'll take a deeper look into how Ctrl+F and others work and then move it into them. I have a feeling that they don't work properly.

File: src/actions/commands/actions.ts [hunk 2]
```diff
@@ -653,13 +653,66 @@ class CommandMoveHalfPageDown extends CommandEditorScroll {
...
+    let editor = vscode.window.activeTextEditor!;
+    let startColumn = vimState.cursorStartPosition.character;
```
Comment (xconverge) [inline src/actions/commands/actions.ts]: can we use const for a lot of these

Comment (rebornix) [reply src/actions/commands/actions.ts]: good suggestion ;)

**Comment by jpoon** [comment]
I'm going to merge this and push a release. We can fix the `let` => `const` later.

**Comment by chiuan** [comment]
still not working ?

**Comment by jpoon** [comment]
Hm, you're right, `ctrl+d` works correctly, but the `ctrl+u` behaviour isn't just right yet. I see it hitting the top of the file.

**Pull request merged**